### DISCUSSION TITLE
Get full request properties

### DISCUSSION THREAD

**Discussion by chrisnoden** [nuxt/nuxt#17317] [Questions] [answered]
To create valid <head> meta tags for social media, SEO, etc, I need to derive the host, method, port and path that has been requested by the web browser.

In a dev context this might be `http://localhost:3000/blog` but in production it could be `https://mywebsite.com/blog`

Even looking server-side at `useRequestEvent` I can't see how to get the hostname used - the method (GET) and path is evident.

I suspect I'll have to derive the host server-side and store in state to re-use it client-side, but so far I've hit a dead end on how to reliably get the host.

**Comment by Diizzayy** [comment, marked as answer]
@chrisnoden You're on the right track with `useRequestEvent`, you can extract the host header can leverage it as needed.

> I suspect I'll have to derive the host server-side and store in state to re-use it client-side, but so far I've hit a dead end on how to reliably get the host.

This can be done if you opt to do so, though you can also use `window.location.host` to get the host / domain in browser.

> To create valid meta tags for social media, SEO, etc, I need to derive the host, method, port and path that has been requested by the web browser.

I have a SEO composable that you should find interesting, it extracts the host and protocol as well as sets SEO defaults as needed

`nuxt-app/composables/seo.ts`
```ts
import { useState, useRoute, useRequestHeaders } from '#imports'

type SEOData = {
  title: string,
  desc?: string,
  image?: string,
  type?: string,
};

export const useSEO = (data : SEOData) => {
  const route = useRoute()

  const domain = ref('https://example.com')

  // use this if you'd like to propagate the domain to the client (for usage outside this script)
  // const domain = useState('_domain', () => 'https://example.com')

  if (process.server) {
    const headers = useRequestHeaders()

    const host = headers?.host
    const protocol = headers?.referer?.split('/')?.[0]

    if (host) { domain.value = protocol ? `${protocol}//${host}` : `https://${host}` }
  }

  const title = data.title ? `${data.title} | Diizzayy` : 'Diizzayy'
  const canonical = `${domain.value}${route.fullPath}`

  const desc = data?.desc || 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.'

  const image = data?.image || `${domain.value}/images/icon.png`

  const link = [
    { rel: 'canonical', href: canonical },
    { rel: 'alternate', href: canonical, hreflang: 'en' }
  ]

  const opengraph = [
    { hid: 'og:url', property: 'og:url', content: canonical },
    { hid: 'og:locale', property: 'og:locale', content: 'en_US' },
    { hid: 'og:type', property: 'og:type', content: data.type || 'website' },
    { hid: 'og:title', property: 'og:title', content: title },
    { hid: 'og:description', property: 'og:description', content: desc },
    { hid: 'og:site_name', property: 'og:site_name', content: 'Diizzayy' },
    { hid: 'og:image:alt', property: 'og:image:alt', content: title },
    { hid: 'og:image', property: 'og:image', content: image }
  ]

  const twitter = [
    { hid: 'twitter:card', property: 'twitter:card', content: 'summary_large_image' },
    { hid: 'twitter:site', property: 'twitter:site', content: '@diizzayy' },
    { hid: 'twitter:creator', property: 'twitter:creator', content: '@diizzayy' },
    { hid: 'twitter:title', property: 'twitter:title', content: title },
    { hid: 'twitter:description', property: 'twitter:description', content: desc },
    { hid: 'twitter:image', property: 'twitter:image', content: image },
    { hid: 'twitter:image:alt', property: 'twitter:image:alt', content: title }
  ]

  useHead({
    title,
    link,
    meta: [...opengraph, ...twitter]
  })
}
```

- Usage
```vue
<script lang="ts" setup>
useSEO({ title: 'Home' })
</script>
```

**Reply by chrisnoden** [reply]
That is an incredibly useful and succinct response. Your code really moves me along. I'll be sure you credit you. You are a star 🌟

Thank you

**Reply by Diizzayy** [reply]
@chrisnoden No problem at all, I'm happy to help.
